### PR TITLE
House eval page

### PR DESCRIPTION
**Pull request by AaronHampshire**
Would love suggestions for how to better design my eval view. There's a lot of redundancy and I'm sure there's a better way to accomplish much of it.

### CODE REVIEW

File: RealEstate/apps/core/templates/core/houseEval.html [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+	<h4><em>{{house.address}}</em></h4>
+	<ul class="list-group">
+		{% for category, score in grades %}
+			<li class="list-group-item">
+				<div class="row">
+					<div class="col-sm-6 vcenter">
+						<h4>{{category.summary}}</h4>
+						<p>
+							{% if category.description %}
+								{{ category.description }}
```
Comment (cdschmitz) [inline RealEstate/apps/core/templates/core/houseEval.html:40]: I would reverse the order of the p tag and the 'if category' check, otherwise there is sometimes an empty p tag for no reason.

File: RealEstate/apps/core/templates/core/houseEval.html [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+							{% endif %}
+						</p>
+					</div>
+					<div class="col-sm-6 vcenter rating">
+						<div class="inner-rating" >
+							<center>
+								<span class="help-text" style="padding-right: 15px">Bad </span>
+									<input id={{category.summary}} class='slide' data-slider-id='slide' type="text" data-slider-min="1" data-slider-max="5" data-slider-step="1" data-slider-value="{% if not score %}3{% else %}{{ score }}{% endif %}"/>
+								<span class="help-text" style="padding-left: 15px">Good</span>
+							</center>
```
Comment (cdschmitz) [inline RealEstate/apps/core/templates/core/houseEval.html:50]: Category.summary is a user controlled field, so using input id={{category.summary}} ties the DOM structure to any value they put there, including categories with spaces/random punctuation that browsers might not support.  Something like "category_{{category.id}}" is safer.

File: RealEstate/apps/core/views.py [hunk 3]
```diff
@@ -45,3 +46,83 @@ def get(self, request, *args, **kwargs):
...
+    """
+    View for the Home Evaluation Page. Currently, this page is decoupled
+    from the rest of the app and uses static elements in the database.
+    """
+    def get(self, request, *args, **kwargs):
+        homebuyer = Homebuyer.objects.filter(user_id=request.user.id)
+        couple = Couple.objects.filter(homebuyer__user=request.user)
+        categories = Category.objects.filter(couple=couple)
+
+        # A specific house should be passed into the request, changing next two lines.
```
Comment (cdschmitz) [inline RealEstate/apps/core/views.py]: Yeah, the URL conf should be something like /eval/{house_id}/.  Then the view will contain checks to make sure that house ID is associated with the homebuyer or realtor from the request, otherwise raising PermissionDenied.

File: RealEstate/apps/core/views.py [hunk 4]
```diff
@@ -45,3 +46,83 @@ def get(self, request, *args, **kwargs):
...
         return render(request, 'core/homebuyerHome.html', {'couple': couple, 'house': house})
+
+
+class EvalView(BaseView):
+    """
+    View for the Home Evaluation Page. Currently, this page is decoupled
+    from the rest of the app and uses static elements in the database.
+    """
+    def get(self, request, *args, **kwargs):
+        homebuyer = Homebuyer.objects.filter(user_id=request.user.id)
```
Comment (cdschmitz) [inline RealEstate/apps/core/views.py]: I added a method to the User object which gives an easier way to the homebuyer.
homebuyer = request.user.role_object

This property will return the Homebuyer instance if the user is a homebuyer, or the Realtor instance if they are a realtor (or None if neither).

File: RealEstate/apps/core/views.py [hunk 5]
```diff
@@ -45,3 +46,83 @@ def get(self, request, *args, **kwargs):
...
+                    missing = False
+                    break
+            if missing:
+                graded.append((category, None))
+                
+        class ContactForm(forms.Form):
+            def __init__(self, *args, **kwargs):
+                super(ContactForm, self).__init__(*args, **kwargs)
+                for c, s in graded:
+                    self.fields[c.summary] = forms.CharField(initial="3" if None else s, widget=forms.HiddenInput())
```
Comment (cdschmitz) [inline RealEstate/apps/core/views.py]: This form is defined twice, once in each of the get and post methods.  I saw the initial value is different but I dont think they should be.  Add to forms.py and import.

File: RealEstate/apps/core/views.py [hunk 6]
```diff
@@ -45,3 +46,83 @@ def get(self, request, *args, **kwargs):
...
+            if missing:
+                graded.append((category, None))
+
+        class ContactForm(forms.Form):
+            def __init__(self, *args, **kwargs):
+                super(ContactForm, self).__init__(*args, **kwargs)
+                for c, s in graded:
+                    self.fields[c.summary] = forms.CharField(initial="0" if None else s, widget=forms.HiddenInput())
+        
+        context = {'couple': couple, 'house' : house[0], 'grades': graded, "form" : ContactForm(), "success": True }
```
Comment (cdschmitz) [inline RealEstate/apps/core/views.py]: Django has a mechanism for flashing one-off messages.

from django.contrib import messages
messages.success(request, 'message here')

Then in the template:
{% boostrap_messages %}

This would be a replacement for the 'success' and bootstrap_alert.

**Comment by cdschmitz** [comment]
General comment about redundancy.  There are a lot of spots, especially in the template, where there are hard coded dependencies on the 1-5 Grade scale.  It would be nice if we could just define this once in the models.py and have it 'just work' if, for example, we decided on a 1-7 scale, or 1-10 scale, or whatever.  In the Grade model, we have a choices attribute on the score field defined as ((1, '1'), (2,'2'),...), but I think we should change this to ((1, 'Poor'), (2, 'Below Average'), ..., (5, 'Excellent')).  We can then use Django introspection to set the values appropriately.

score_field = Grade._meta.get_field('score')
score_field.default    # 3
score_field.choices   #  Defined above

From the choices field we can calculate: 
- The minimum and maximum values that the score field can have ( for data-slider-min and data-slider-max).
-  A dictionary that maps values to their verbose representation (ex:  {1: 'Poor'}).  This dict can be converted to a string and injected directly into the js script as an object, so the slider function is just simple lookup based on value.

File: RealEstate/apps/core/views.py (hunk 5 above)
Comment (AaronHampshire) [inline RealEstate/apps/core/views.py]: @cdschmitz: Totally agree. However, looking down the road, this whole save/reload thing seems really 2005. There's also hacky issues here where I'm creating a form of hidden fields and binding those fields to the sliders via javascript in order to take advantage of the form posting abilities. However, most of this can be simplified using ajax. 

Maybe next sprint, we can include a smaller extension story to work ajax in? (I can see a similar issue with the category page, too.) And if we do, I'd rather not pull out forms and then forget they've been pulled. 

What do you think?

Comment (cdschmitz) [inline RealEstate/apps/core/views.py]: I agree that things would ideally be async, however my main concern is the time considerations since we are already a few weeks into the second term.  So I think the highest priority should be something that works, then we can revisit things to improve them.  The one merge-blocking issue for this pull request is the URLconf, which should include the house ID as described in a separate comment.

Comment (AaronHampshire) [inline RealEstate/apps/core/views.py]: I agree with you. I think I got a bit tunnel visioned into this feature. I'll push up changes later tonight.
